presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -235,6 +239,54 @@ public RowExpression combineDisjunctsWithDefault(Collection<RowExpression> expre
         return disjuncts.isEmpty() ? emptyDefault : or(disjuncts);
     }
 
+    /**
+     * Given a logical expression, the goal is to push negation to the leaf nodes.
+     * This only applies to propositional logic. this utility cannot be applied to high-order logic.
+     * Examples of non-applicable cases could be f(a AND b) > 5
+     *
+     * An applicable example:
+     *
+     *        NOT
+     *         |
+     *      ___OR_                          AND
+     *     /      \                      /      \
+     *    NOT     OR        ==>        AND      AND
+     *     |     /  \                 /  \     /   \
+     *    AND   c   NOT              a    b   NOT  d
+     *   /  \        |                         |
+     *  a    b       d                         c
+     */
+    public RowExpression pushNegationToLeaves(RowExpression expression)
+    {
+        if (isNotCallExpression(expression)) {
+            RowExpression argument = ((CallExpression) expression).getArguments().get(0);
+            // eliminate two consecutive negations
+            if (isNotCallExpression(argument)) {
+                return pushNegationToLeaves(((CallExpression) argument).getArguments().get(0));
+            }
+            // push negation through conjunction or disjunction
+            if (isConjunctionOrDisjunction(argument)) {
+                SpecialFormExpression specialForm = ((SpecialFormExpression) argument);
+                if (specialForm.getForm() == AND) {
+                    // !(a AND b) ==> !a OR !b
+                    return or(pushNegationToLeaves(notCallExpression(specialForm.getArguments().get(0))), pushNegationToLeaves(notCallExpression(specialForm.getArguments().get(1))));
```

Comment:
nit: I in general prefer not to static import things like `and` / `or`  :). But it's up to you . 
Also maybe consider reorg as
```java
        return or(
                pushNegationToLeaves(
                        notCallExpression(specialForm.getArguments().get(0))),
                pushNegationToLeaves(
                        notCallExpression(specialForm.getArguments().get(1))));
```

This might be a bit easier to parse by human ? :)